components/service/experiments/src/main/java/mozilla/components/service/experiments/ExperimentsUpdater.kt

```diff
@@ -0,0 +1,110 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.experiments
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import androidx.work.*
+import java.util.concurrent.TimeUnit
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.GlobalScope
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * Class used to schedule sync of experiment
+ * configuration from the server using WorkManager
+ */
+internal class ExperimentsUpdater(
+    private val context: Context,
+    private val experiments: ExperimentsInternalAPI
+) {
+    private val logger: Logger = Logger(LOG_TAG)
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var source: ExperimentSource? = null
+
+    inner class SyncWorker(context: Context, workerParams: WorkerParameters) : androidx.work.Worker(context, workerParams) {
+        override fun doWork(): Result {
+            return if (updateExperiments()) {
+                Result.success()
+            } else {
+                Result.retry()
+            }
+        }
+    }
+
+    fun initialize(configuration: Configuration) {
+        if (source == null) {
+            source = KintoExperimentSource(
+                EXPERIMENTS_BASE_URL,
+                EXPERIMENTS_BUCKET_NAME,
+                EXPERIMENTS_COLLECTION_NAME,
+                configuration.httpClient.value
+            )
+        }
+
+        GlobalScope.launch(IO) {
+            // Update the experiments list from the server async, without blocking
+            // the app launch.
+            updateExperiments()
+
+            // Schedule future updates of the experiments list every few hours.
+            schedule()
+        }
+    }
+
+    /**
+     * Schedule updating the experiments regularly.
+     *
+     * @param worker worker class
+     */
+    fun schedule() {
+        val constraints = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+        /*
+        val syncWork = PeriodicWorkRequest.Builder(SyncWorker::class.java, UPDATE_INTERVAL_HOURS, TimeUnit.HOURS)
+            .addTag(TAG)
+            .setConstraints(constraints)
+            .build()
+        WorkManager.getInstance().enqueueUniquePeriodicWork(TAG, ExistingPeriodicWorkPolicy.KEEP, syncWork)
+        */
+        WorkManager.getInstance().enqueue(OneTimeWorkRequest.from(SyncWorker::class.java))
```

Comment:
@Dexterp37  To simplify, let's just use a one-time request, to highlight the issue.

Running this from the Glean sample results in a run-time error:

> 2019-03-27 18:14:28.245 27026-27044/org.mozilla.samples.glean.debug E/WM-WorkerFactory: Could not instantiate mozilla.components.service.experiments.ExperimentsUpdater$SyncWorker
>     java.lang.NoSuchMethodException: <init> [class android.content.Context, class androidx.work.WorkerParameters]
>         at java.lang.Class.getConstructor0(Class.java:2327)
>         at java.lang.Class.getDeclaredConstructor(Class.java:2166)
>         at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:91)
>         at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:233)
>         at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:127)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
>         at java.lang.Thread.run(Thread.java:764)
> 2019-03-27 18:14:28.245 27026-27044/org.mozilla.samples.glean.debug E/WM-WorkerWrapper: Could not create Worker mozilla.components.service.experiments.ExperimentsUpdater$SyncWorker
>

Reply:
My assumption was... that `SyncWorker`s constructor is not visible to the factory/workmanager code.

Reply:
Gah, you're perfectly right, and I just had a closer look at glean's code: it's erroring out there as well.

Reply:
Ok, I dug into this a bit more. We can only really pass strings and other basic types to a `Worker` via [`Data`](https://developer.android.com/reference/androidx/work/Data.html) (as @travis79 knows). In order to have more complex injection in place, we'd need to provide the `WorkManager` with a custom [`WorkerFactory`](https://developer.android.com/reference/androidx/work/WorkerFactory). This would allow us to instantiate Workers however we want.

**However**, in order to enable a custom `WorkerFactory` we'd need to disable the automatic `WorkManager` init in the (app? library?) manifest and [manually init](https://developer.android.com/reference/androidx/work/WorkManager.html#initialize(android.content.Context,%20androidx.work.Configuration)) it by passing our custom configuration object with the custom `WorkerFactory`.

**Problem** I'm not sure we can do that from the library, and I'm not sure we should ask consumers to manually disable the `WorkManager` init from the library. These are uncharted waters for me, @pocmo might know better. The fallback option is to do what you didn't want to, calling static APIs.

Reply:
Disabling the default WorkManager init may have other consequences if other bits of the client app are using WorkManager.  I'm not sure that's a safe bet as I don't know even if we did it from the library it wouldn't have some effect on the client app.

I think if this is going to work with the way we have it structured now, then we need to see if there's a way to serialize/reconstruct the object from the Worker class.  This worked well with the PingUploadWorker serializing pings and just reading them from disk.  Is there any way to create a data class around the bits you need for the Worker to function, and then use the [`@Parcelize`](https://github.com/Kotlin/KEEP/blob/master/proposals/extensions/android-parcelable.md) functionality to pass it to the Worker?

Reply:
Yeah, the work manager can't create an instance of your class since it is an inner class, which means it can only be created inside `ExperimentsUpdater` ... because it keeps a reference to its outer class - and that's why you can call updateExperiments() in the outer class.

Here's what I suggest to do:
* Create just a normal class that extends Worker.
* Inside `doWork()` you can access your `Experiments` singleton and from there you should get to your `updateExperiments()` (or add a way to get there).

There's no need to pass anything to the worker here...?

Reply:
@pocmo I was wondering how to decouple this from the singleton for unit testing.
But maybe i should just live with this for now?

Reply:
@georgf Yeah, singletons are always a pain in tests. It's one of the reasons why in other components we leave creating the singleton to the app (with the help of a DI framework if the app chooses so).

For tests in other situations we tried to hide the singleton part behind a method that we can mock in a test (e.g. if we can't mock System.currentTimeMillis() then we hide this behind a local now() method that we can mock).

Reply:
Interesting, do you have a link to such a function and its mocks in tests handy? :)

Reply:
@georgf I can point you at an example like he's talking about: [this](https://github.com/mozilla-mobile/android-components/blob/761f43462120913f2d9e0e00c8d73e37dd9a214d/components/service/telemetry/src/main/java/org/mozilla/telemetry/schedule/jobscheduler/TelemetryJobService.java#L139)

It's mocked in the `testDailyLimitIsEnforced()` test [here](https://github.com/mozilla-mobile/android-components/blob/210751ad2b02639cbfc729b0e96a0c2dc204e846/components/service/telemetry/src/test/java/org/mozilla/telemetry/schedule/jobscheduler/TelemetryJobServiceTest.java#L72)

The Kotlin version should be pretty close to that.  I think we have some examples of mocking functions using the `when` syntax in service-glean already, such as [here](https://github.com/mozilla-mobile/android-components/blob/1c017bf3d87b4ed738bd5c9a2f809cc5ce2b0ee8/components/service/glean/src/test/java/mozilla/components/service/glean/TestUtil.kt#L155)

Reply:
Cheers!